src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs

```diff
@@ -0,0 +1,122 @@
+{
+    static internal List<Func<Task>> CleanupMethods { get; set; } = new List<Func<Task>>();
+
+    public override TestResult[] Execute(ITestMethod testMethod)
+    {
+        var timeout = PlaywrightSettingsProvider.TestTimeout;
+        if (timeout == null || Debugger.IsAttached)
+        {
+            return base.Execute(testMethod);
+        }
+        if (PlaywrightTimeoutIsSmallerThanMSTestTimeout(testMethod, timeout.Value, out var result))
+        {
+            return result!;
+        }
+
+        return ExecuteWithTimeoutAsync(testMethod, timeout.Value).GetAwaiter().GetResult();
+    }
+
+    private async Task<TestResult[]> ExecuteWithTimeoutAsync(ITestMethod testMethod, int timeout)
+    {
+        // 1. Race Test execution against the specified timeout.
+        var testExecutionTask = Task.Run(() => base.Execute(testMethod));
+        var timeoutTask = Task.Delay(timeout);
+        await Task.WhenAny(testExecutionTask, timeoutTask);
+
+        // 2. If timeout was reached, call the tearDownHandlers and
+        // cancel by that the underlying BrowserContext calls.
+        if (timeoutTask.IsCompleted)
+        {
+            // 2.1 Call the test teardown handlers.
+            try
+            {
+                foreach (var cleanupMethod in CleanupMethods)
+                {
+                    await cleanupMethod();
+                }
+            }
+            finally
+            {
+                CleanupMethods.Clear();
+            }
+
+            // 2.2 Wait for the test execution to complete (let e.g. Page.ClickAsync throw)
+            // and give the test 5 seconds to finish.
+            await Task.WhenAny(testExecutionTask, Task.Delay(5000));
+        }
+
+        if (testExecutionTask.IsCompleted)
+        {
+            return await testExecutionTask;
+        }
+
+        return new TestResult[]
+        {
+            new TestResult()
+            {
+                Outcome = UnitTestOutcome.Timeout,
+                TestFailureException = new Exception($"Test timed out after {timeout}ms.")
+            },
+        };
+    }
+
+    private bool PlaywrightTimeoutIsSmallerThanMSTestTimeout(ITestMethod testMethod, int playwrightTestTimeout, out TestResult[]? result)
+    {
+        var timeout = MSTestProvider.TestTimeout;
+        var timeoutAttribute = testMethod.MethodInfo.GetCustomAttributes(typeof(TimeoutAttribute), false).FirstOrDefault();
+        if (timeoutAttribute != null)
+        {
+            timeout = ((TimeoutAttribute)timeoutAttribute).Timeout;
+        }
+
+        if (timeout != null && timeout != 0 && playwrightTestTimeout > timeout)
+        {
+            result = new TestResult[]
+            {
+                new TestResult()
+                {
+                    Outcome = UnitTestOutcome.Failed,
+                    TestFailureException = new Exception($"Playwright test timeout ({playwrightTestTimeout}ms) is larger than MSTest timeout ({timeout}ms).")
```

Comment:
Is it smaller or larger? How should it be? I find naming here somewhat confusing.

Reply:
I can rephrase it to e.g. `Playwright.TestTimeout is larger than MSTest.TestTimeout, make sure to have a larger Playwright timeout otherwise test execution gets terminated`